platooning_strategic_IHP/src/platoon_strategic_ihp.cpp

```diff
@@ -814,8 +1038,123 @@ namespace platoon_strategic_ihp
         }
     }
 
+    // UCLA: Mobility operation callback for lead_with_operation state (cut-in join).
+    void PlatoonStrategicIHPPlugin::mob_op_cb_leadwithoperation(const cav_msgs::MobilityOperation& msg)
+    {
+        std::string strategyParams = msg.strategy_params;
+        bool isPlatoonStatusMsg = (strategyParams.rfind(OPERATION_STATUS_TYPE, 0) == 0);
+        if(isPlatoonStatusMsg) 
+        {
+            mob_op_cb_STATUS(msg);
+            ROS_DEBUG_STREAM("Platoon STATUS operation message processed in Lead with Operation state.");
+        } 
+        else 
+        {
+            ROS_DEBUG_STREAM("Received a mobility operation message with params " << msg.strategy_params << " but ignored.");
+        }
+    }
+
+    // UCLA: Mobility operation callback for prepare to join state (cut-in join).
+    void PlatoonStrategicIHPPlugin::mob_op_cb_preparetojoin(const cav_msgs::MobilityOperation& msg)
+    {
+        /*
+         * If same lane with leader, then send request to do same lane join. 
+         * Otherwise, just send status params.
+         * Note: leader in state 'leading with operation' also send out INFO msg
+         */
+
+        // read parameters
+        std::string strategyParams = msg.strategy_params;
+        std::string senderId = msg.header.sender_id; 
+        std::string platoonId = msg.header.plan_id;  
+        std::string newLeaderID = msg.header.sender_id; 
+        std::string newLeaderPlatoonId = pm_.currentPlatoonID; 
+        // locate INFO type
+        bool isPlatoonInfoMsg = (strategyParams.rfind(OPERATION_INFO_TYPE, 0) == 0);
+        bool isPlatoonStatusMsg = (strategyParams.rfind(OPERATION_STATUS_TYPE, 0) == 0);
+        bool isNotInNegotiation = (pm_.current_plan.valid == false);
+
+        if (isPlatoonInfoMsg)
+        {
+            // read ecef location from strategy params.
+            cav_msgs::LocationECEF ecef_loc;
+            ecef_loc = mob_op_find_ecef_from_INFO_params(strategyParams);
+            
+            // use ecef_loc to calculate front Dtd in m.
+            lanelet::BasicPoint2d incoming_pose = ecef_to_map_point(ecef_loc);
+            double frontVehicleDtd = wm_->routeTrackPos(incoming_pose).downtrack;
+
+            // use ecef_loc to calculate front Ctd in m.
+            double frontVehicleCtd = wm_->routeTrackPos(incoming_pose).crosstrack;
+
+            // Use pm_ to find platoon end vehicle and its downtrack in m.
+            int rearVehicleIndex = pm_.platoon.size() - 1;
+            double rearVehicleDtd = pm_.platoon[rearVehicleIndex].vehiclePosition; 
+            
+            // downtrack of the platoon leader --> used for frontal join
+            ROS_DEBUG_STREAM("rearVehicleDtd from ecef: " << rearVehicleDtd);
+
+            // determine joining vehicle position 
+            bool isSameLaneWithPlatoon = (frontVehicleCtd - current_crosstrack_ > -1 || frontVehicleCtd - current_crosstrack_ < 1);
```

Comment:
Why not just use isVehicleNextLane()?  This is slightly different logic than in there, but it seems it should all be the same.  If it really needs to be different, it would be worth a comment here explaining why.